File: class22/java/PES2202-Microkernel.java
/*
 * PES2202 - Microkernel (Plugin Architecture)
 * Aula 22: Microsserviços e Arquiteturas Distribuídas
 * Projeto e Engenharia de Software - Prof. Giovane Barcelos
 *
 * Implementa um core mínimo (MicrokernelCore) que não conhece nenhum
 * plugin concreto, apenas o contrato IExportPlugin. Novos formatos de
 * exportação (PDF, CSV, ...) entram como plugins registrados em tempo
 * de execução, sem alterar uma linha do core.
 *
 * Compilar/executar:
 *   javac PES2202-Microkernel.java && java Microkernel
 */

import java.util.*;

// CONTRATO que todo plugin de exportação deve implementar
interface IExportPlugin {
    String exportar(Map<String, String> dados);
}

class PluginPDF implements IExportPlugin {
    @Override
    public String exportar(Map<String, String> dados) {
        return "[PDF] Exportando " + dados.size() + " registros";
    }
}

class PluginCSV implements IExportPlugin {
    @Override
    public String exportar(Map<String, String> dados) {
        StringBuilder sb = new StringBuilder();
        for (Map.Entry<String, String> e : dados.entrySet()) {
            if (sb.length() > 0) sb.append(";");
            sb.append(e.getKey()).append("=").append(e.getValue());
        }
        return "[CSV] " + sb;
    }
}

// CORE: não conhece PluginPDF nem PluginCSV, só o contrato IExportPlugin
class MicrokernelCore {
    private final Map<String, IExportPlugin> plugins = new HashMap<>();

    void registrarPlugin(String nome, IExportPlugin plugin) {
        plugins.put(nome, plugin);
        System.out.println("[Core] Plugin '" + nome + "' registrado.");
    }

    String exportarCom(String nome, Map<String, String> dados) {
        IExportPlugin plugin = plugins.get(nome);
        if (plugin == null) {
            throw new IllegalArgumentException("Plugin '" + nome + "' não registrado");
        }
        return plugin.exportar(dados);
    }
}

class Microkernel {
    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("  MICROKERNEL (PLUGIN) — EXPORTAÇÃO DE DADOS");
        System.out.println("=".repeat(60));

        MicrokernelCore core = new MicrokernelCore();
        core.registrarPlugin("pdf", new PluginPDF());
        core.registrarPlugin("csv", new PluginCSV());

        Map<String, String> dados = new LinkedHashMap<>();
        dados.put("nome", "Fusca");
        dados.put("placa", "ABC1234");

        System.out.println("\n" + core.exportarCom("pdf", dados));
        System.out.println(core.exportarCom("csv", dados));

        try {
            core.exportarCom("xml", dados);
        } catch (IllegalArgumentException e) {
            System.out.println("\n[Esperado] " + e.getMessage());
        }

        System.out.println("\n✓ Core nunca importa PluginPDF/PluginCSV diretamente, só IExportPlugin.");
        System.out.println("✓ Adicionar um novo formato (ex.: XML) não exige alterar o core.");
    }
}

File: class22/python/PES2202-Microkernel.py
"""
PES2202 - Microkernel (Plugin Architecture)
Aula 22: Microsserviços e Arquiteturas Distribuídas
Projeto e Engenharia de Software - Prof. Giovane Barcelos

Implementa um core mínimo (MicrokernelCore) que não conhece nenhum
plugin concreto, apenas o contrato IExportPlugin. Novos formatos de
exportação (PDF, CSV, ...) entram como plugins registrados em tempo
de execução, sem alterar uma linha do core.
"""

from abc import ABC, abstractmethod


# CONTRATO que todo plugin de exportação deve implementar
class IExportPlugin(ABC):
    @abstractmethod
    def exportar(self, dados: dict) -> str:
        pass


class PluginPDF(IExportPlugin):
    def exportar(self, dados: dict) -> str:
        return f"[PDF] Exportando {len(dados)} registros"


class PluginCSV(IExportPlugin):
    def exportar(self, dados: dict) -> str:
        linhas = ";".join(f"{k}={v}" for k, v in dados.items())
        return f"[CSV] {linhas}"


# CORE: não conhece PluginPDF nem PluginCSV, só o contrato IExportPlugin
class MicrokernelCore:
    def __init__(self):
        self._plugins: dict[str, IExportPlugin] = {}

    def registrar_plugin(self, nome: str, plugin: IExportPlugin) -> None:
        self._plugins[nome] = plugin
        print(f"[Core] Plugin '{nome}' registrado.")

    def exportar_com(self, nome: str, dados: dict) -> str:
        if nome not in self._plugins:
            raise ValueError(f"Plugin '{nome}' não registrado")
        return self._plugins[nome].exportar(dados)


if __name__ == "__main__":
    print("=" * 60)
    print("  MICROKERNEL (PLUGIN) — EXPORTAÇÃO DE DADOS")
    print("=" * 60)

    core = MicrokernelCore()
    core.registrar_plugin("pdf", PluginPDF())
    core.registrar_plugin("csv", PluginCSV())

    dados = {"nome": "Fusca", "placa": "ABC1234"}
    print("\n" + core.exportar_com("pdf", dados))
    print(core.exportar_com("csv", dados))

    try:
        core.exportar_com("xml", dados)
    except ValueError as e:
        print(f"\n[Esperado] {e}")

    print("\n✓ Core nunca importa PluginPDF/PluginCSV diretamente, só IExportPlugin.")
    print("✓ Adicionar um novo formato (ex.: XML) não exige alterar o core.")
